include/revng/Pipeline/ReadPathsCache.h

```diff
@@ -0,0 +1,215 @@
+  }
+
+  Target &getTarget() { return Content; }
+
+  std::string &getStepName() { return StepName; }
+
+  std::string &getContainerName() { return ContainerName; }
+
+  bool operator==(const LocatedTarget &) const = default;
+
+  bool operator!=(const LocatedTarget &) const = default;
+
+  bool operator<(const LocatedTarget &Other) const {
+    const auto &LHS = std::tie(Content, StepName, ContainerName);
+    return LHS < std::tie(Other.Content, Other.StepName, Other.ContainerName);
+  }
+
+  ParsableLocatedTarget toParsable() const;
+};
+
+class TargetReadPathsSerializable {
+public:
+  std::string Path;
+};
+
+class PathSerializationPrimitives {
+public:
+  PathSerializationPrimitives() = default;
+
+  template<TupleTreeCompatibleAndVerifiable T>
+  static PathSerializationPrimitives make() {
+    PathSerializationPrimitives Primitives;
+    Primitives.Deserialize = &stringAsPath<T>;
+    Primitives.Serialize = &pathAsString<T>;
+    return Primitives;
+  }
+
+  std::optional<TupleTreePath> (*Deserialize)(llvm::StringRef);
+  std::optional<std::string> (*Serialize)(const TupleTreePath &);
+};
+
+class TargetReadPaths {
+private:
+  TupleTreePath Path;
+
+public:
+  explicit TargetReadPaths(const TupleTreePath &Path) : Path(Path) {}
+
+  static llvm::Expected<TargetReadPaths>
+  fromParsable(const Context &Ctx,
+               const TargetReadPathsSerializable &Target,
+               PathSerializationPrimitives Primitives);
+
+  const TupleTreePath &getPaths() const { return Path; }
+
+  TupleTreePath &getPaths() { return Path; }
+
+  bool operator==(const TargetReadPaths &Other) const = default;
+  bool operator!=(const TargetReadPaths &Other) const = default;
+
+  bool operator<(const TargetReadPaths &Other) const {
+    return std::tie(Path) < std::tie(Other.Path);
+  }
+
+  TargetReadPathsSerializable
+  toParsable(PathSerializationPrimitives Primitives) const;
+};
+
+class SerializableReadPathCache {
+public:
+  using ValueType = std::pair<pipeline::TargetReadPathsSerializable,
+                              pipeline::ParsableLocatedTarget>;
+  using Vector = std::vector<ValueType>;
+  Vector Data;
+};
+
+class ReadPathCache {
+private:
+  using MapType = std::map<TargetReadPaths, std::vector<LocatedTarget>>;
+  using ReverseMapType = std::map<LocatedTarget, std::vector<TargetReadPaths>>;
+  MapType Map;
+  ReverseMapType ReverseMap;
+  PathSerializationPrimitives Primitives;
+
+public:
+  explicit ReadPathCache(PathSerializationPrimitives Primitives) :
+    Map(), Primitives(Primitives) {}
+
+  void clear() { Map = MapType(); }
+
+  template<TupleTreeCompatibleAndVerifiable T>
+  static ReadPathCache make() {
+    return ReadPathCache(PathSerializationPrimitives::make<T>());
+  }
+
+  void insert(const LocatedTarget &LocatedTarget,
+              const TargetReadPaths &TargetReadPaths) {
+    Map[TargetReadPaths].push_back(LocatedTarget);
+    ReverseMap[LocatedTarget].push_back(TargetReadPaths);
```

Comment:
Can you elaborate on why we need both directions for the map?
It's not obvious to me.

Reply:
we need the one from path to target so that we can figure out who is invalidated when the element pointed by that path changes, when we need the other to be able to compute the hash when we are doing the cold start, as well as to check if a target has been produced without committing it

Reply:
Please add a `TODO` stating that we can remove one of the two directions of the map once the tracking becomes mandatory ("strict").

For the cold start use case, we can revert the map on demand.